src/activations.jl

```diff
@@ -75,7 +76,7 @@ Leaky [Rectified Linear Unit](https://en.wikipedia.org/wiki/Rectifier_(neural_ne
 activation function.
 You can also specify the coefficient explicitly, e.g. `leakyrelu(x, 0.01)`.
 """
-leakyrelu(x, a=oftf(x, 0.01)) = max(a * x, x)
+leakyrelu(x, a=oftf(x, 0.01)) = ifelse(x>0, float(x), oftf(x, a*x))  # max(a*x, x) is 3x slower
```

Comment:
https://github.com/FluxML/NNlib.jl/pull/345/commits/e6befa18f6ce5d1d71d776159eb7b43a3120253e is really orthogonal, does this:
```julia
# Unrelated, improve relu slightly:
julia> x = randn(Float32, 10^3); y = similar(x);
julia> relu_max(x) = max(zero(x), x);
julia> relu_if(x) = ifelse(x>0, x, zero(x));
julia> x = randn(Float32, 10^3); y = similar(x);
julia> @btime $y .= relu_max.($x);
  110.752 ns (0 allocations: 0 bytes)
julia> @btime $y .= relu_if.($x);
  92.354 ns (0 allocations: 0 bytes)
  
# and improve leakyrelu quite a bit
julia> oftf(x, y) = oftype(float(x), y);
julia> leakyrelu_max(x, a=oftf(x, 0.01)) = max(a * x, x);
julia> leakyrelu_if(x, a=oftf(x, 0.01)) = ifelse(x>0, float(x), oftf(x, a*x))
leakyrelu_if (generic function with 2 methods)
julia> @btime $y .= leakyrelu_max.($x);
  558.396 ns (0 allocations: 0 bytes)
julia> @btime $y .= leakyrelu_if.($x);
  164.928 ns (0 allocations: 0 bytes)
  ```

Reply:
How does perf look without requiring `float(x)`? e.g. for mixed, lower or higher precision.

Reply:
I put `float(x)` there to match the other branch, as `ifelse` doesn't promote, but `max` does. Without the float, `@code_warntype leakyrelu_if(1)` is bad.

Both seem type-stable with anything I could think of, e.g. `@code_warntype leakyrelu_if(Float16(1), 0.3)`; I'm not entirely sure why `leakyrelu_max` is so much slower.

Reply:
Right, my point was this means `typeof(leakyrelu(Float16(1))) == typeof(leakyrelu(Float64(1)))`. Namely one gets promoted and the other reduced.

Reply:
No, it preserves Float16:
```
julia> typeof(leakyrelu_if(Float16(1))) == typeof(leakyrelu_if(Float64(1)))
false

julia> typeof(leakyrelu_max(Float16(1))) == typeof(leakyrelu_max(Float64(1)))
false

julia> leakyrelu_if(Float16(1))
Float16(1.0)

julia> leakyrelu_if(Float16(-1))
Float16(-0.01)

julia> leakyrelu_if(Float16(1), 3//4)
Float16(1.0)

julia> leakyrelu_if(Float16(-1), 3//4)
Float16(-0.75)
```

Reply:
I wasn't aware `float` didn't promote `Float_` types, interesting. As for the lowering, it carries over into llvm as well:
```julia
julia> @code_llvm leakyrelu_if(Float16(1))
;  @ REPL[12]:1 within `leakyrelu_if`
define half @julia_leakyrelu_if_403(half %0) #0 {
top:
;  @ REPL[12]:1 within `leakyrelu_if` @ REPL[12]:1
; ┌ @ operators.jl:378 within `>`
; │┌ @ float.jl:511 within `<` @ float.jl:444
    %1 = fpext half %0 to float
    %2 = fcmp ule float %1, 0.000000e+00
; └└
; ┌ @ float.jl:403 within `*`
   %3 = fmul float %1, 0x3F847C0000000000
   %4 = fptrunc float %3 to half
; └
  %5 = select i1 %2, half %4, half %0
;  @ REPL[12]:1 within `leakyrelu_if`
  ret half %5
}

julia> @code_llvm leakyrelu_if(Float32(1))
;  @ REPL[12]:1 within `leakyrelu_if`
define float @julia_leakyrelu_if_412(float %0) #0 {
top:
;  @ REPL[12]:1 within `leakyrelu_if` @ REPL[12]:1
; ┌ @ operators.jl:378 within `>`
; │┌ @ float.jl:490 within `<` @ float.jl:443
    %1 = fcmp ule float %0, 0.000000e+00
; └└
; ┌ @ float.jl:404 within `*`
   %2 = fmul float %0, 0x3F847AE140000000
; └
  %3 = select i1 %1, float %2, float %0
;  @ REPL[12]:1 within `leakyrelu_if`
  ret float %3
}
```
As well as mentioned in https://github.com/JuliaLang/julia/blob/bb5b98e72a151c41471d8cc14cacb495d647fb7f/src/llvm-demote-float16.cpp#L38.

Using `@fastmath` saves on one truncation:
```julia
julia> leakyrelu_if_fast(x, a=oftf(x, 0.01)) = @fastmath ifelse(x>0, float(x), oftf(x, a*x));

julia> @code_llvm leakyrelu_if_fast(Float16(1))
;  @ REPL[15]:1 within `leakyrelu_if_fast`
define half @julia_leakyrelu_if_fast_378(half %0) #0 {
top:
;  @ REPL[15]:1 within `leakyrelu_if_fast` @ REPL[15]:1
; ┌ @ operators.jl:378 within `>`
; │┌ @ float.jl:511 within `<` @ float.jl:444
    %1 = fpext half %0 to float
    %2 = fcmp ule float %1, 0.000000e+00
; └└
; ┌ @ fastmath.jl:167 within `mul_fast`
   %3 = fmul fast half %0, 0xH211F
; └
  %4 = select i1 %2, half %3, half %0
;  @ REPL[15]:1 within `leakyrelu_if_fast`
  ret half %4
}

julia> @code_llvm leakyrelu_if_fast(Float32(1))
;  @ REPL[15]:1 within `leakyrelu_if_fast`
define float @julia_leakyrelu_if_fast_380(float %0) #0 {
top:
;  @ REPL[15]:1 within `leakyrelu_if_fast` @ REPL[15]:1
; ┌ @ operators.jl:378 within `>`
; │┌ @ float.jl:490 within `<` @ float.jl:443
    %1 = fcmp ule float %0, 0.000000e+00
; └└
; ┌ @ fastmath.jl:167 within `mul_fast`
   %2 = fmul fast float %0, 0x3F847AE140000000
; └
  %3 = select i1 %1, float %2, float %0
;  @ REPL[15]:1 within `leakyrelu_if_fast`
  ret float %3
}
```

Reply:
Interesting. Both are super-slow on Float16 though. 
```
julia> x = randn(Float16, 10^3); y = similar(x);

julia> @btime $y .= leakyrelu_max.($x);
  29.583 μs (0 allocations: 0 bytes)

julia> @btime $y .= leakyrelu_if.($x);
  7.708 μs (0 allocations: 0 bytes)

julia> @btime $y .= leakyrelu_if_fast.($x);
  7.729 μs (0 allocations: 0 bytes)
```

Reply:
Makes sense to me, the native asm (at least on x86) generated is far more complex for the Float16 path. Probably something to do with hardware support.

Reply:
Moved to #347 now.